infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java

```diff
@@ -29,19 +44,36 @@
 @Slf4j
 public class LogEntryWriter {
     private final LogReplicationMetadataManager logReplicationMetadataManager;
-    private final HashMap<UUID, String> streamMap; //the set of streams that log entry writer will work on.
+    private final LogReplicationConfig config;
+    private final CorfuRuntime rt;
+
+    // When the streams to replicate are dynamically discovered, there is no guarantee that
+    // during Log Entry sync the stream being replicated was already registered to registry table.
+    // Those streams' replicated entries need to be added to this pending entries map for streaming
+    // consideration, as the stream tags are fetched from registry table.
+    private final Map<UUID, List<SMREntry>> pendingEntries;
+
+    // Set of streams newly opened and replicated from ACTIVE, need to clear local
+    // writes to avoid polluting those streams.
+    private final Set<UUID> streamsToClear;
+    private Set<UUID> streamsSet; //the set of streams that log entry writer will work on.
     private long srcGlobalSnapshot; //the source snapshot that the transaction logs are based
     private long lastMsgTs; //the timestamp of the last message processed.
-    private Map<UUID, List<UUID>> dataStreamToTagsMap;
+    private Map<UUID, Set<UUID>> dataStreamToTagsMap;
+    private boolean handlePendingEntries; // Flag to indicate LogEntryWriter to handle pending entries
 
-    public LogEntryWriter(LogReplicationConfig config, LogReplicationMetadataManager logReplicationMetadataManager) {
+    public LogEntryWriter(LogReplicationConfig config, LogReplicationMetadataManager logReplicationMetadataManager,
```

Comment:
I have a question, as far as I remember, registryTable updates are also replicated (mergeOnlyStreams). Can you point me where in the code you check if it's already present so it's not overwritten with incorrect info (as source and sink might be on different versions?)

Reply:
The registryTable updates won't be filtered out during log replications, since I thought even on a rolling upgrade case, we cannot tell whether registryTable have correct info or not.

The correctness is now guarded by the registryTable's own update protocol, as long as standby tried to open a table again (after log replication) and the CorfuRecord of that table is different with that in registryTable, registryTable will update that entry.

And that's also part of the reason we need to get noisy stream set before log replication starts to filter out those explicitly set `is_federated` to false